web-services-app/src/main/java/edu/unc/lib/boxc/web/services/processing/IiifV3ManifestService.java

```diff
@@ -0,0 +1,241 @@
+import edu.unc.lib.boxc.search.api.models.Datastream;
+import edu.unc.lib.boxc.web.common.services.AccessCopiesService;
+import info.freelibrary.iiif.presentation.v3.AnnotationPage;
+import info.freelibrary.iiif.presentation.v3.Canvas;
+import info.freelibrary.iiif.presentation.v3.ImageContent;
+import info.freelibrary.iiif.presentation.v3.Manifest;
+import info.freelibrary.iiif.presentation.v3.PaintingAnnotation;
+import info.freelibrary.iiif.presentation.v3.properties.Label;
+import info.freelibrary.iiif.presentation.v3.properties.Metadata;
+import info.freelibrary.iiif.presentation.v3.properties.RequiredStatement;
+import info.freelibrary.iiif.presentation.v3.services.ImageService3;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static edu.unc.lib.boxc.model.api.DatastreamType.JP2_ACCESS_COPY;
+
+/**
+ * Service for generating iiif v3 manifests for repository object
+ * @author bbpennel
+ */
+public class IiifV3ManifestService {
+    private static final Logger log = LoggerFactory.getLogger(IiifV3ManifestService.class);
+    private AccessCopiesService accessCopiesService;
+    private AccessControlService accessControlService;
+    private String baseIiifv3Path;
+    private String baseAccessPath;
+    private String baseServicesApiPath;
+
+    /**
+     * Constructs a manifest record for the object identified by the provided PID
+     * @param pid
+     * @param agent
+     * @return
+     */
+    public Manifest buildManifest(PID pid, AgentPrincipals agent) {
+        assertHasAccess(pid, agent);
+        var contentObjs = accessCopiesService.listViewableFiles(pid, agent.getPrincipals());
+        if (contentObjs.size() == 0) {
+            throw new NotFoundException("No objects were found for inclusion in manifest for object " + pid.getId());
+        }
+        log.debug("Constructing manifest for {} containing {} items", pid.getId(), contentObjs.size());
+        ContentObjectRecord rootObj = contentObjs.get(0);
+        var manifest = new Manifest(getManifestPath(rootObj), new Label(getTitle(rootObj)));
+        manifest.setMetadata(buildMetadata(rootObj));
+        addAttribution(manifest, rootObj);
+
+        addItems(manifest, contentObjs);
+
+        return manifest;
+    }
+
+    private List<Metadata> buildMetadata(ContentObjectRecord rootObj) {
+        var metadataList = new ArrayList<Metadata>();
+        String abstractText = rootObj.getAbstractText();
+        if (abstractText != null) {
+            metadataList.add(new Metadata("description", abstractText));
+        }
+        addMetadataField(metadataList, "Creators", rootObj.getCreator());
+        addMetadataField(metadataList, "Subjects", rootObj.getSubject());
+        addMetadataField(metadataList, "Languages", rootObj.getLanguage());
+        metadataList.add(new Metadata("", "<a href=\"" +
```

Comment:
why does this one not use the `addMetadataField` method?

Reply:
the method has some list joining stuff happening for non-null values which doesn't apply to this or the description field. I'll update the method name to make it clearer